incydr/cli/cmds/users.py

```diff
@@ -239,33 +264,127 @@ def move(ctx: Context, user, org_guid):
-    Requires the following columns:
+     By default, the provided roles will replace the specified user's existing roles.
+     Use the --add flag or the --remove flag to add or remove roles, respectively, from a user's existing roles.
+
+    Takes a single arg FILE which specifies the path to the file.  Use the --format option to specify the format of your input file.  Defaults to csv.
+
+    Requires the following CSV columns or JSON keys:
 
     * `user` - User ID or username of the user whose roles will be updated. Performs an additional lookup if username is passed.
-    * `roles` - Space-delimited list of role IDs to assign to the new user.  These will replace the specified user's existing roles.
+    * `role` - Role ID and/or role name (case-sensitive) to assign to the new user.
+
+    Example:
+
+        The following command will add roles to users as specified in a JSON-lines formatted file:
+
+            incydr users bulk-update-roles path/to/file.json --add --format json-lines
     """
+
+    @lru_cache()
+    def resolve_username(user):
+        if user is None:
+            return
+        elif "@" in user:
+            return user_lookup(client, user)
+        else:  # assume user_id
+            return user
+
+    class RoleUpdateCSV(CSVModel):
+        user: str = Field(csv_aliases=["user", "user_id", "username", "id", "userId"])
+        role: str = Field(
+            csv_aliases=["role", "role_id", "role_name", "roleId", "roleName"]
+        )
+
+    class RoleUpdateJSON(Model):
+        user: str
+        role: str
+
+    if format_ == "csv":
+        roles_file = RoleUpdateCSV.parse_csv(file)
+    else:
+        roles_file = RoleUpdateJSON.parse_json_lines(file)
+
     client = ctx.obj()
-    try:
-        for row in track(
-            read_dict_from_csv(csv),
-            description="Updated user roles...",
-            transient=True,
-        ):
-            user = row.get("user")
-            if user and "@" in user:
-                user = user_lookup(client, row["user"])
-            client.users.v1.update_roles(user, row["roles"].split(" "))
-    except HTTPError as err:
-        console.print(f"[red]Error:[/red] {err.response.text}")
+    if update_method == "add":
+        update_func = client.users.v1.add_roles
+    elif update_method == "remove":
+        update_func = client.users.v1.remove_roles
+    else:
+        update_func = client.users.v1.update_roles
+
+    # group updates by user
+    buckets = bucketize(
+        roles_file,
+        key=lambda role_update: role_update.user,
+        value_transform=lambda role_update: role_update.role,
+    )
+    for bucket in buckets:
+        user = bucket
+        roles = buckets[bucket]
+        try:
+            user = resolve_username(user)
+            update_func(user, roles)
+            console.print(f"Successfully updated roles for user {user}.'")
+        except RoleProcessingError as err:
```

Comment:
I've changed this to raise a single error that wraps a list of all errors that arose during role processing.  I'm not sure if there's any `HTTPErrors` that would appropriate to handle here? FWIW If you try and access/update the roles with an incorrect ID it returns a `403 forbidden` which could be raised for a number of other reasons.